src/libcore/array.rs

```diff
@@ -285,3 +304,171 @@ macro_rules! array_impl_default {
 }
 
 array_impl_default!{32, T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T}
+
+
+/// An iterator that moves out of an array.
+///
+/// This `struct` is created by the `into_iter` method on [arrays]
+/// (provided by the [`IntoIterator`] trait).
+///
+/// [arrays]: ../../std/primitive.array.html
+/// [`IntoIterator`]: ../../std/iter/trait.IntoIterator.html
+#[unstable(feature = "array_into_iter", issue = "0")]
+pub struct IntoIter<T, A: FixedSizeArray<T>> {
+    // Invariants: index <= index_back <= array.len()
+    // Only values in array[index..index_back] are alive at any given time.
+    // Values from array[..index] and array[index_back..] are already moved/dropped.
+    array: ManuallyDrop<A>,
+    index: usize,
```

Comment:
One thing I've personally enjoyed doing historically is storing these two indices as `Range<usize>` as it helps make implementations for the iterator methods easier